wp-includes/sqlite/class-wp-sqlite-pdo-engine.php

```diff
@@ -385,97 +378,148 @@ private function prepare_directory() {
+				return $this->commit();
+			}
+			if ( preg_match( '/^ROLLBACK/i', $statement ) ) {
+				return $this->rollBack();
+			}
 
-		$this->queries[] = "Raw query:\n$statement";
-		$res             = $this->determine_query_type( $statement );
-		if ( ! $res && defined( 'PDO_DEBUG' ) && PDO_DEBUG ) {
-			$bailout_string = sprintf(
-				'<h1>Unknown query type</h1><p>Sorry, we cannot determine the type of query that is requested (%s).</p>',
-				$statement
-			);
-			$this->set_error( __LINE__, __FUNCTION__, $bailout_string );
-		}
-		switch ( strtolower( $this->query_type ) ) {
-			case 'set':
-				$this->return_value = false;
-				break;
-
-			case 'foundrows':
-				$_column = array( 'FOUND_ROWS()' => '' );
-				$column  = array();
-				if ( ! is_null( $this->found_rows_result ) ) {
-					$this->num_rows          = $this->found_rows_result;
-					$_column['FOUND_ROWS()'] = $this->num_rows;
-					$column[]                = new WP_SQLite_Object_Array( $_column );
-					$this->results           = $column;
-					$this->found_rows_result = null;
-				}
-				break;
+			$this->translator = new WP_SQLite_Translator( $this->pdo, $GLOBALS['table_prefix'] );
 
-			case 'insert':
-				if ( $this->can_insert_multiple_rows ) {
-					$this->execute_insert_query_new( $statement );
-					break;
+			do {
+				$error = null;
+				try {
+					$translation = $this->translator->translate(
+						$statement,
+						$this->found_rows_result
+					);
+				} catch ( PDOException $error ) {
+					if ( $error->getCode() !== self::SQLITE_BUSY ) {
+						return $this->handle_error( $error );
+					}
 				}
-				$this->execute_insert_query( $statement );
-				break;
-
-			case 'create':
-				$this->return_value = $this->execute_create_query( $statement );
-				break;
-
-			case 'alter':
-				$this->return_value = $this->execute_alter_query( $statement );
-				break;
-
-			case 'show_variables':
-				$this->return_value = $this->show_variables_workaround( $statement );
-				break;
-
-			case 'showstatus':
-				$this->return_value = $this->show_status_workaround( $statement );
-				break;
-
-			case 'drop_index':
-				$this->return_value = false;
-				if ( preg_match( '/^\\s*(DROP\\s*INDEX\\s*.*?)\\s*ON\\s*(.*)/im', $statement, $match ) ) {
-					$this->query_type   = 'alter';
-					$this->return_value = $this->execute_alter_query( 'ALTER TABLE ' . trim( $match[2] ) . ' ' . trim( $match[1] ) );
+			} while ( $error );
+
+			$stmt        = null;
+			$last_retval = null;
+			foreach ( $translation->queries as $query ) {
+				$this->queries[] = "Executing: {$query->sql} | " . ( $query->params ? 'parameters: ' . implode( ', ', $query->params ) : '(no parameters)' );
+				do {
+					$error = null;
+					try {
+						$stmt        = $this->pdo->prepare( $query->sql );
+						$last_retval = $stmt->execute( $query->params );
+					} catch ( PDOException $error ) {
+						if ( $error->getCode() !== self::SQLITE_BUSY ) {
+							throw $error;
+						}
+					}
+				} while ( $error );
+			}
```

Comment:
This entire block should be inside of a transaction. There might be one active already, though, so `beginTransaction()` and friends should support nested transactions via `SAVEPOINT` and `RELEASE`